hooks/keydata.go

```diff
@@ -0,0 +1,284 @@
+		}
+		ciphertext = aead.Seal(nil, nonce, payload, aad)
+
+		var encryptedKey []byte
+		encryptedKey, handle, err = keyProtector.ProtectKey(rand, symKey, nil)
+		if err != nil {
+			return nil, nil, nil, fmt.Errorf("cannot protect symmetric key for AEAD compat: %w", err)
+		}
+
+		aeadCompat = &aeadCompatData{
+			Nonce:        nonce,
+			EncryptedKey: encryptedKey,
+		}
+	default:
+		ciphertext, handle, err = keyProtector.ProtectKey(rand, payload, aad)
+		if err != nil {
+			return nil, nil, nil, fmt.Errorf("cannot protect key: %w", err)
+		}
+	}
+
+	kd, err := secbootNewKeyData(&secboot.KeyParams{
+		Handle: &KeyData{
+			data: keyData{
+				Handle:     handle,
+				Scope:      *scope,
+				AEADCompat: aeadCompat,
+			},
+		},
+		Role:             params.Role,
+		EncryptedPayload: ciphertext,
+		PlatformName:     platformName,
+		KDFAlg:           kdfAlg,
+	})
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("cannot create key data: %w", err)
+	}
+
+	return kd, primaryKey, unlockKey, nil
+}
+
+// NewKeyData creates a new KeyData object for the supplied secboot.KeyData.
+func NewKeyData(k *secboot.KeyData) (*KeyData, error) {
+	var kd *KeyData
+	if err := k.UnmarshalPlatformHandle(&kd); err != nil {
+		return nil, fmt.Errorf("cannot unmarshal platform key data: %w", err)
+	}
+	kd.k = k
+
+	return kd, nil
+}
+
+// MarshalJSON implements [json.Marshaler].
+func (d KeyData) MarshalJSON() ([]byte, error) {
+	return json.Marshal(d.data)
+}
+
+// UnmarshalJSON implements [json.Unmarshaler].
+func (d *KeyData) UnmarshalJSON(data []byte) error {
+	var kd keyData
+	if err := json.Unmarshal(data, &kd); err != nil {
+		return err
+	}
+	d.data = kd
+	return nil
+}
+
+// SetAuthorizedSnapModels updates the snap models that this key data is
+// bound to. This update has to be authenticated, which requires the primary
+// key that was originally supplied to [NewProtectedKey].
+//
+// On success, this will automatically update the corresponding *[secboot.KeyData]
+// that this key data was created from using [NewKeyData].
+func (d *KeyData) SetAuthorizedSnapModels(key secboot.PrimaryKey, models ...secboot.SnapModel) error {
+	if err := d.data.Scope.SetAuthorizedSnapModels(key, d.k.Role(), models...); err != nil {
+		return err
+	}
+	return d.k.MarshalAndUpdatePlatformHandle(d)
+}
+
+// SetAuthorizedBootModes updates the boot modes that this key data is
+// bound to. This update has to be authenticated, which requires the primary
+// key that was originally supplied to [NewProtectedKey].
+//
+// On success, this will automatically update the corresponding *[secboot.KeyData]
+// that this key data was created from using [NewKeyData].
+func (d *KeyData) SetAuthorizedBootModes(key secboot.PrimaryKey, modes ...string) error {
+	if err := d.data.Scope.SetAuthorizedBootModes(key, d.k.Role(), modes...); err != nil {
+		return err
+	}
+	return d.k.MarshalAndUpdatePlatformHandle(d)
+}
```

Comment:
these methods don't seem to be exercised in tests

Reply:
I've added some tests for both of these methods now.